### PR TITLE
Add ABORT_ON_WASM_EXCEPTIONS

### PR DESCRIPTION
**Pull request by tklajnscek**
Putting this up for discussion - this PR adds `Module['onException']` as discussed here:
https://github.com/emscripten-core/emscripten/issues/11667

It can be used as a catch-all notification for unhandled exceptions that occur within the execution frame of Emscripten compiled code.

`onException` fires when an exception occurs:
- in main function
- in the registered mainloop function
- in any code invoked via ccall/cwrap/embind

The exception is still passed on, because we need the execution of the JS task to get interrupted.

### CODE REVIEW

**Comment by kripken** [comment]
I have three concerns with this: first, it adds to code size. We could avoid that by making it optional somehow (at the cost of some complexity).

The second concern is that there are potentially a lot of code paths that need to be covered here, so it seems risky for bugs. For example non-ccall calls are not handled here. Instead, how about instrumenting all wasm exports? See for example `Asyncify.instrumentWasmExports` which instruments them for another reason. That would guarantee that any call to an export is handled.

My third concern is that I'd like to avoid complexity in the codebase. I think instrumenting all exports as just mentioned would help there, as this would be all in a single place. In fact, I think this could be done using the existing `Module['instantiateWasm']` hook (see preamble.js), so maybe we don't need a new option, and can just document how to do it with that? I'm not sure though.

**Comment by tklajnscek** [comment]
Nice one! It works fine if I instrument all the calls, but it's a huge pain in the ass to do it via `instantiateWasm` because we have to manually roll all the instantiation code and I'd prefer to keep the default Emscripten code with the sync fallbacks etc as they are.

I think it's fine if we hook into this via a `preRun` callback. This fires before the runtime initializes so it's early enough that everything should pick up the instrumented calls from `Module.asm`, including embind bindings that initialize as global objects.

I guess the same reasons then apply to https://github.com/emscripten-core/emscripten/pull/11694 which would need to instrument the calls like this. I could just add this to my wrappers, but unfortunately 'var ABORT' is not exposed on the Module so would it make sense to have something that lets external code check it's value? Something like `Module.hasAborted()`?

Then I can pretty much do everything we need with a simple wrapper like so:
```
preRun: (Module) => {
  var asm = {};
  for (var name in Module['asm']) {
    (function(name) { // Create a closure (for loop semantics)
      var original = Module['asm'][name];
      if (typeof original === 'function' && name != 'main') { // Skip main because it's already guarded etc
        asm[name] = function() {
          if (Module.hasAborted()) { // There's no way to check this at the moment
            throw "the program has already aborted!";
          }
          try {
            return original.apply(null, arguments);
          }
          catch (e) {
            abort("An unhandled exception in '" + name + "' has potentially corrupted the program state. Halting execution!");
          }
        };
      }
      else {
        asm[name] = original;
      }
    })(name);
  }
  Module['asm'] = asm;
}
```

**Comment by tklajnscek** [comment]
I still think it would be useful to add the stuff in this PR and https://github.com/emscripten-core/emscripten/pull/11694 in some way because it should just be super easy to:
- get notified your program crashed
- stop calls into the program from working after that

There's simply no way to make that reliable once you've bailed out mid execution.
People will get bitten by this, waste a bunch of time on it and hopefully find some of these discussions.

If nothing else, we could at least put some warnings somewhere in portability docs about this:
- the program may abort or throw but it can still be called into via exported functions which is likely to fails
- there's no way of easily catching the crashes unless you manually wrap every call in a try/catch block (we can include an example)

Just for reference, here's the final form of the epic crash handler that seems to do everything I need it to do right now:
```
/**
 * Abort and unhandled exception handling
 *
 * This makes the Emscripten program behave more like a proper native program by making sure 
 * that all hooks are unregistered immediately when it crashes so no further scheduled or 
 * manual calls into the program happen. 
 *
 * We instrument all the exported functions so we can catch crashes in the scheduled mainloop/step
 * function and any API calls in addition to main() that's handled already.
 *
 * After the program crashes none of the functions are allowed to be called and throw
 * and exception if they are.
 *
 * There's two ways to crash the program:
 * 1) calling abort() which invokes the onAbort handler that performs cleanup
 * 2) throwing an exception which just gets passed into abort() so that the flow is the same
 */
Module['preRun'] = function() {
  // First add an abort handler that cleans up everything when an abort happens
  var prevAbort = Module["onAbort"];
  Module["onAbort"] = function(err) {
    _postCrashCleanup(); 
    if(prevAbort) {
      prevAbort(err);
    }
  };
  
  // Then instrument all the exported functions to:
  // - abort if an unhandled exception occurs
  // - throw an exception if someone tries to call them after the program has aborted
  // Exclude main since it's handled by postamble.js:callMain already.
  var asm = {};
  var mainFunc = Module["asm"]["main"];
  for (var name in Module['asm']) {
    (function(name) { 
      var original = Module['asm'][name];
      if (typeof original === 'function' && original != mainFunc) {
        asm[name] = function() {
          // Don't allow this function to be called if we're aborted!
          if (ABORT) { 
            throw "the program has already aborted!";
          }
          
          try {
            return original.apply(null, arguments);
          }
          catch (e) {
            // abort() might've been called in the original function call above and we caught it's exception
            // If so, just rethrow it.
            if(ABORT) {
              throw e;
            }

            abort("unhandled exception: " + (typeof e === 'object' && e.stack) ? e.stack : e);
          }
        };
      }
      else {
        asm[name] = original;
      }
    })(name);
  }
  Module['asm'] = asm; 
}
```

**Comment by kripken** [comment]
Interesting! I worry that `preRun` is not quite right, as it has no guarantee on `Module['asm']` being loaded yet. There may be a race condition there. `preMain` is the right thing here I believe.

I think I agree there may be value to add something like that. How about adding checks in ASSERTIONS mode, and for the non-debugging stuff, it could be a new option? Perhaps `HANDLE_WASM_EXCEPTIONS` but hopefully we can find a better name. Another way to make this optional would be to check for `Module.onAbort` in `INCOMING_MODULE_JS_API` - that's included by default in the list there, but at least it would be easy to remove for people that don't need it (and it would be off for MINIMAL_RUNTIME by default).

**Comment by tklajnscek** [comment]
> I worry that `preRun` is not quite right, as it has no guarantee on `Module['asm']` being loaded yet. There may be a race condition there. `preMain` is the right thing here I believe.

I think `preMain` is too late, because `initRuntime` runs before it which runs the static initializers, which is where embind reigsters everything and caches functions before we've instrumented them. It seems the best place is to actually run this in a `__ATINIT__ `callback so it will run before `___wasm_call_ctors`.



> I think I agree there may be value to add something like that. How about adding checks in ASSERTIONS mode, and for the non-debugging stuff, it could be a new option?

Yeah, that seems like a good idea. I think something like ABORT_ON_UNHANDLED_EXCEPTION would work and is very descriptive?

I'll push some changes for review soon.

**Comment by sbc100** [comment]
We already have createExportWrapper that is used to wrap exports in debug mode:

https://github.com/emscripten-core/emscripten/blob/df5f6d150a4206b6935346b51fbd84da664c4b58/src/preamble.js#L804

Would it make sense to piggyback on that mechanism?

**Comment by tklajnscek** [comment]
> We already have createExportWrapper that is used to wrap exports in debug mode:
> 
> https://github.com/emscripten-core/emscripten/blob/df5f6d150a4206b6935346b51fbd84da664c4b58/src/preamble.js#L804
> 
> Would it make sense to piggyback on that mechanism?

Yeah, thanks for the suggestion, that will probably work fine! Especially if I made it emit a stripped down version of that when ASSERTIONS==0 and ABORT_ON_WASM_EXCEPTIONS==1. I'll have some time to look at this tomorrow and update the issue.

**Comment by tklajnscek** [comment]
OK so after spending a full day on this, here's the status.

TL;DR: Added a new setting ABORT_ON_WASM_EXCEPTIONS that puts a function into __ATINIT__ that instruments all exports with a try/catch block, aborts when an exception is throw and doesn't allow calling into the code anymore after an abort has happened. It adds ~250 bytes in optimized builds, but it's a fixed overhead regardless of how many exports there are.

I also had a version that used `createExportWrapper` as @sbc100 suggested, which was super easy for ASSERTIONS builds. Then I made it emit the wrapper for optimized builds which caused closure to fail so I had to do some juggling to make it work with the delay_assignment (basically, had a function getter for the function that was being wrapped so that we could keep the same 'Module['asm']['name'] literal that closure properly minified. Unfortunately wrapping it this way caused the dead code elimination to fail and it always kept all the exports which added quite a bit to the size of optimized builds. In the end I went for the standalone wrapper that's much cleaner and self contained.

Here's the other variant, for reference:

File: src/preamble.js [hunk 1]
```diff
@@ -481,6 +481,58 @@ function preRun() {
   callRuntimeCallbacks(__ATPRERUN__);
 }
 
+#if ABORT_ON_WASM_EXCEPTIONS
+// Instrument all the exported functions to:
+// - abort if an unhandled exception occurs
+// - throw an exception if someone tries to call them after the program has aborted
+// Set settings.ABORT_ON_WASM_EXCEPTIONS for more info.
+__ATINIT__.push(function exceptionInstrument() {
```
Comment (sbc100) [inline src/preamble.js:489]: Why not `addOnInit()`?  Or at last unshift so that it will be the first one run?

Comment (tklajnscek) [reply src/preamble.js:489]: Whoops, that was .unshift at some point. I was following the way wasm_ctors gets added which is directly into __ATINIT__ hence no addOnInit(). I'm fine either way.

**Comment by tklajnscek** [comment]
> I don't really mind using a different method to wrap the exports but if we do I think we should use it for the assert checks too. I'd like to avoid having two different methods of wrapping.

Yeah totally agree with that, but didn't want to make too big a change. I'll see if I can find some time to attempt a unified solution based on my code.

> I also don't really understand the problem you described with createExportWrapper.. but I do remember that code being extremely sensitive. If its can be done purely in JS and not need the python code that seems better.

Basically doing this in optimized builds:
`var _someFunc = Module["_someFunc"] = createExportWrapper("someFunc", function() { return asm["someFunc"]; });`
made all of these not be eliminated whereas this (the way it used to be in optimized builds) did get eliminated:
`return (_someFunc = Module["_someFunc"] = Module["asm"]["someFunc"]).apply(null, arguments);`

Couldn't figure out why and ran out of time fighting what seemed like an uphill battle.

**Comment by tklajnscek** [comment]
I also need to check what's killing the tests 🤭

**Comment by tklajnscek** [comment]
The instrumenting approach had a bunch of issues, especially with exception handling and the fact that the wrappers were being run within wrappers etc so I backtracked and decided to figure out what was killing dead code removal. Turns out the dead code removal is a bit hardcodey 😅 

So I updated `acorn-optimizer.js` to actually dead-code-eliminate this:
`var _someFunc = Module["_someFunc"] = createExportWrapper(*);`
when `_someFunc` is unused and it works great.

This now actually produces smaller optimized JS files with ABORT_ON_WASM_EXCEPTIONS because it avoids repeating `.apply(null, arguments);` for every generated export 😎 

I've tried a subset of the tests locally (including the C++ exception tests etc.) and they pass fine. When CircleCI starts working again we'll see whether they all pass or not as I can't run all of them locally (getting some errors, not failures; might be due to being on windows or the env setup)

**Comment by kripken** [comment]
I think adding multiple changes like this:
```
var _someFunc = Module["_someFunc"] = createExportWrapper(*);
```
is less efficient than iterating on the exports object as `instrumentWasmExports` does,

https://github.com/emscripten-core/emscripten/blob/8715a1bb4a6b614e3de6b842f8e57e054c2ba265/src/library_async.js#L318-325

That should also avoid needing any acorn changes I hope?

**Comment by sbc100** [comment]
> I think adding multiple changes like this:
> 
> ```
> var _someFunc = Module["_someFunc"] = createExportWrapper(*);
> ```
> 
> is less efficient than iterating on the exports object as `instrumentWasmExports` does,

But this change doesn't add those does it?  createExportWrapper is a pre-existing thing.

See the DECLARE_ASM_MODULE_EXPORTS setting for if/when/why we sometime choose to declare these exports like this directly on the global object rather than in a loop.

Or am I misunderstanding your comment here?

> 
> https://github.com/emscripten-core/emscripten/blob/8715a1bb4a6b614e3de6b842f8e57e054c2ba265/src/library_async.js#L318-325
> 
> That should also avoid needing any acorn changes I hope?

**Comment by sbc100** [comment]
I'd be all in favor of removing DECLARE_ASM_MODULE_EXPORTS if there is any way we can.

**Comment by kripken** [comment]
@sbc100 

> createExportWrapper is a pre-existing thing.

Oh, I didn't know that, sorry.

I do think it would be a good improvement to do before this PR, then - move that into an instrumentation on the exports, like asyncify does. Or is there a reason not to do so? If we still need access to the original exports as in dynamic linking that may be an issue, but otherwise wrapping the exports is simpler, safer, and smaller I think.

**Comment by sbc100** [comment]
I think its tricky.    I recently changed the way we wrap exported to insert the debug assertions which is where ` createExportWrapper` was born.    And I think it relates to the fact that in `DECLARE_ASM_MODULE_EXPORTS` mode we need to create these top level objects at JS load time before the wasm module is even loaded.

Can you explain what you don't like about the `createExportWrapper` appoach?  I mean doesn't it do basically the same thing.. it essentially wraps all the exports.

**Comment by kripken** [comment]
> Can you explain what you don't like about the createExportWrapper appoach?

It's more complicated: we need to emit those calls and also identify them properly in some JS optimizer passes, which need to recognize export receiving (unless we don't care about those optimizations, if these wrappers are just for debugging). It also increases code size (one call per export). And it means code in emscripten.py as well as in the JS, I think? (the other approach needs to no emscripten.py changes)

In general it's simpler to just have a single place that loops on the exports and wraps them, I think, the way Asyncify does. That is,
```javascript
var exports = ...
exports = instrumentExports(exports);
..
function instrumentExports(exports) {
  loop on
    export = wrapping(export);
}
```
as opposed to
```javascript
var export1 = wrapping(export1);
var export2 = wrapping(export2);
..
var exportN = wrapping(exportN);
```

If we are sure these wrappings are never important in optimized builds, then I'm not strongly opposed, though, in that case it's mostly bikeshedding.

**Comment by sbc100** [comment]
I agree with you in principle I just seem to remember this be in conflict with DECLARE_ASM_MODULE_EXPORTS which means we have to declare each of the exported top level vars at the top level.   Do they have to at least be declared i the form:

```
var export1 = ..something..
```

But perhaps there is something we put there as a placeholder that instrumentExports can then update.

I think one problem is that once `export1` is bound at the top level we don't want to rebind it.   If somebody captured `export1` before the module was loaded that export1 would need to mutate into something that works rather than be replaced (I think).

**Comment by tklajnscek** [comment]
I agree with @kripken that it would make more sense and be better for code size if the exports were simply registered from JS. But it ain't that simple... here's my take on this after looking at this for a while, reading various discussions and comments in the code.

First, createExportWrapper is currently only there to aid in debugging and give the user nice error messages (runtime not initialized, function not found etc.) and doesn't even work in Closure optimized builds it's current state because it uses the unminified name to find look up the function. So this is debug only.

Second, WASM & WASM_ASYNC_COMPILATION are what everyone should be targeting by default so most people will be hitting the `delay_assignment==true` path in `create_receiving` which emits this for every export:
```
var _someFunc = Module["_someFunc "] = function() {
  return (_someFunc = Module["_someFunc"] = Module["asm"]["someFunc"]).apply(null, arguments);
};
```
or after minification something like this:
```
var a=b._someFunc=function(){return(a=b._someFunc=e.asm.c).apply(null,arguments)};
// The export asm["someFunc"] gets minified too via MINIFY_WASM_IMPORTS_AND_EXPORTS
```
Some observations about the above:
- The user cannot store the value of `var _someFunc` or `Module["_someFunc"]` because they change on first use and they'll keep calling the wrong function if they do. So every call needs to go through `_someFunc/Module["_someFunc"]` anyways and we can safely reassign them to something else later like the current code does.
- If the user calls the function too soon, before WASM has loaded, it's going to fail because it won't be there yet (or I might be missing something?)
- Even after minification it's still quite lengthy for a _per-export_ overhead, but it does work nicely with Closure which should outweigh this

Also, the current `DECLARE_ASM_MODULE_EXPORTS=0` path has some major issues that pretty much make it unusable:
1. It creates the symbols on the global scope which completely breaks `MODULARIZE=1` - one instance will work (unless something uses the same name on the global scope already), but multiple won't because they will overwrite each other's functions. And it pollutes the global scope. For this to work it would have to create the the symbols in the function's scope, which can't be done as far as I can tell.
2. It will not work with Closure because closure needs to see the symbols declared (e.g. `var someFunc`) explicitly. It will also not work with the Binaryen export minification pass because it won't see `asm["someFunc"]` explicitly but `asm[name]` in a loop. I'm guessing the whole point of having `var someFunc` was to allow internal calls to get minified, otherwise everything would just use `Module['someFunc']` everywhere which would also make it work with `DECLARE_ASM_MODULE_EXPORTS=0` nicely, even with `MODULARIZE=1`, but at the expense of code size. Oh and `MINIMAL_RUNTIME=1` doesn't even have `Module` anymore... 😄 

Btw, completely unintentionally, with `ABORT_ON_WASM_EXCEPTIONS=1` the code size overhead per-export is actually reduced:
```
var a=b._someFunc=function(){return(a=b._someFunc=e.asm.c).apply(null,arguments)};
```
becomes slightly shorter:
```
var a=b._someFunc=d("_someFunc",function(){return e.asm.c});

// 'd' is the createExportWrapper function
// also, this is why asm['_someFunc] is accessed through the function passed into it - so that it can get minified
```

So it seems to stay in the correct scope and keep minifiers happy we need to do most of the stuff as it is now:
- declare `var _someFunc` explicitly in that scope as there's no other way to create the symbols in the correct scope after the fact
- use `Module["asm"]["someFunc"]` explicitly instead of `Module["asm"][name]` in a loop -  so it can get minified 

But if I'm not missing something obvious - since the the exports (`asm["someFunc"]`) get minified anyways why don't we just change all the internal calls to use `asm["someFunc"]`, instead of `someFunc`, then we can get away with a loop that just registers them on Module like so:
```
var exportNames = ["libFunc", "someFunc"];
var exportNamesMin = ["a", "b"]; // We need this from the binaryen export minifier pass
for(var i=0;i<exportNames.length;i++) {
  Module[exportNames[i]] = asm[exportNamesMin[i]];
}
```
and for MINIMAL_RUNTIME it just goes away completely.

Thoughts?

**TL;DR** - Holy crap that was WAY longer than I expected, no TL;DR for this one 🤷

**Comment by kripken** [comment]
@tklajnscek 

>  So this is debug only.

I see, that sounds good then. Sounds ok to me to keep using it for those purposes.

However, I think `ABORT_ON_WASM_EXCEPTIONS` is intended for non-debug builds too?

> DECLARE_ASM_MODULE_EXPORTS=0 path has some major issues that pretty much make it unusable:     It creates the symbols on the global scope which completely breaks MODULARIZE=1 -[only]  one instance will work 

Looks like the docs in settings.js do mention it is dangerous, at least.

> why don't we just change all the internal calls to use asm["someFunc"], instead of someFunc

The former will be minified to something like `q.someFunc` as it can't minify the string name. That can be much larger than the latter which could be minified to a single character. That's really the main issue.

Maybe this is not as much of a problem with import/export name minification, which we have now, and could get to `q.r` at least. Still 3 characters, but if it saves in other places, that could be good. However I think we may need to generalize the import/export name minification for this to work.

But yes, if we can get there, and we think 3 characters is ok, then it does sound good to switch to a loop on the exports - that could be great!

**Comment by tklajnscek** [comment]
> However, I think `ABORT_ON_WASM_EXCEPTIONS` is intended for non-debug builds too?

Yes, that's why I changed `createExportWrapper` a bit to be non-debug friendly and play nice with the minifier and also updated that custom dead code eliminator to include this case. The wrapper that gets emitted in release is minimal as it excludes allthe ASSERTIONS=1 things. On top of that it produces the smallest (currently) possible emitted exports size (see below).

> Maybe this is not as much of a problem with import/export name minification, which we have now, and could get to `q.r` at least. Still 3 characters, but if it saves in other places, that could be good. However I think we may need to generalize the import/export name minification for this to work.

Yeah exactly, it's on by default now.

So just to recap the current situation with the latest Emscripten for both `WASM=1` and `WASM=0 => WASM2JS`:
``` 
// With `-O0`:
var _malloc = Module["_malloc"] = function() {
  return (_malloc = Module["_malloc"] = Module["asm"]["malloc"]).apply(null, arguments);
};

// With `-O3/-O2` the export name is minified (`Module["asm"]["i"]`):
var _malloc=Module["_malloc"]=function(){return(_malloc=Module["_malloc"]=Module["asm"]["i"]).apply(null,arguments)};

// With `-O2 --closure 1` everything but the name on the module (which can't be) is minified:
var V=a._malloc=function(){return(V=a._malloc=a.asm.i).apply(null,arguments)};

// Finally, with `-O2 --closure 1` and with my changes:
var V=a._malloc=E("malloc",function(){return a.asm.i});
```

So with this PR we get the shortest output with the current `var _func+ asm['func']` approach. Granted it's at the expense of an extra function indirection, but it actually does useful work and at least I think it's worth it. And it can be disabled if someone doesn't need it - and they should really only disable this if they handle it themselves, not like now where it's just not handled and can wreak havoc - _Oh but my code never crashes I don't need this_ yeah right 🤣 

The way I see it is if we decide to do a cleanup pass on how the exports are emitted we can easily switch it over to loop on the exports, but without that the loop would produce larger code than it does now, because the per-emitted-export size would be larger.

Any further thoughts on what's keeping us going ahead and merging this @kripken @sbc100 ?

PS: I can possibly spend some time in a week or two seeing what it would take to get rid of the per-export `var _someFunc` by just using asm['someFunc'] everywhere if there's interest in making this change.

**Comment by sbc100** [comment]
> PS: I can possibly spend some time in a week or two seeing what it would take to get rid of the per-export `var _someFunc` by just using asm['someFunc'] everywhere if there's interest in making this change.

I imagine that people will depend on this feature though, no?  Isn't that kind of the whole point of EXPORTED_FUNCTIONS?  Or is exporting on the Module object the correct way to access exported these days?

**Comment by kripken** [comment]
@sbc100 
> I imagine that people will depend on this feature though, no? 

Yes, in general I think we can't remove the `var _someFunc` things. The `DECLARE_ASM_MODULE_EXPORTS` is what disables that, I don't think we can turn it on by default.

We could btw make the JS optimizer smarter and able to remove unused `var _someFunc` declarations, perhaps.

@tklajnscek 
> Yes, that's why I changed createExportWrapper a bit to be non-debug friendly and play nice with the minifier and also updated that custom dead code eliminator to include this case. 

I still think it would be simpler to do it the other way. That is, leave `createExportWrapper` to be something only for debugging. And we can use an asyncify-like wrapping at runtime approach for this. That would avoid any new complexity in the minifier or the dead code eliminator.

**Comment by tklajnscek** [comment]
> I still think it would be simpler to do it the other way. That is, leave `createExportWrapper` to be something only for debugging. And we can use an asyncify-like wrapping at runtime approach for this. That would avoid any new complexity in the minifier or the dead code eliminator.

Sure, I have both versions done. If @sbc100 agrees that's the way I'll put that one up in this PR. Let me know and I'll put it up tomorrow morning, it's late here ;)

**Comment by sbc100** [comment]
> We could btw make the JS optimizer smarter and able to remove unused `var _someFunc` declarations, perhaps.

These should not exist should they?   IIUC `var _someFunc` should only exist for explictly exports functions and they need to stick around so that a user can use them.

**Comment by kripken** [comment]
@sbc100 

> These should not exist should they? IIUC var _someFunc should only exist for explictly exports functions and they need to stick around so that a user can use them.

True, yeah. We could only do this in cases like MODULARIZE I guess, where we know the entire global scope. You're right, it would only be useful for unused exports, which seems rare.

Btw @sbc100 in case you missed it, please respond to @tklajnscek 's question in the last comment - if you agree then we can move forward here.

**Comment by sbc100** [comment]
> > I still think it would be simpler to do it the other way. That is, leave `createExportWrapper` to be something only for debugging. And we can use an asyncify-like wrapping at runtime approach for this. That would avoid any new complexity in the minifier or the dead code eliminator.
> 
> Sure, I have both versions done. If @sbc100 agrees that's the way I'll put that one up in this PR. Let me know and I'll put it up tomorrow morning, it's late here ;)

Fine with me.    I mostly care about not adding yet another wrapping mechanism.  I don't mind which of the existing ones you extent.

**Comment by tklajnscek** [comment]
I've now pushed the asm wrapper version. And added a test for the feature as well.

Anyone have any idea how to make the circleci tests run? I think they only ran on the first commit?

**Comment by sbc100** [comment]
Hmm.. circleci tests should just work.  I can't see why they wouldn't.  Can you try rebasing and pushing again?

**Comment by tklajnscek** [comment]
Did it. No dice.
It only runs CI/archive, the rest never execute :(

**Comment by sbc100** [comment]
Right thats because  CI/archive is not run on circleci but on github actions.   But the circleci jobs *should* be running.

Can you try uploading a dummy/test PR?

**Comment by tklajnscek** [comment]
Put up https://github.com/emscripten-core/emscripten/pull/11934 from a fresh branch.
Seems to be running them now. I remember it ran them here on the first commit, then it stopped.

**Comment by tklajnscek** [comment]
Looking at test run at https://github.com/emscripten-core/emscripten/pull/11934
It seems that this change breaks _dlsym because it wraps the wasm function in a JS function which needs `sig` passing into `addFunction` but none of the code using it ever passes in a signature. Any hints would be greatly appreciated :)

**Comment by tklajnscek** [comment]
Closing this. See #11934
